posts/2018-08-16-well-typed-heap.lhs

```diff
@@ -0,0 +1,1100 @@
+programming downright infeasible due to the sheer amount of lemmas that need to
+be proven.
+
+Consider the relative elegance of defining a type family for incrementing a
+binary number that is read from the right to the left:
+
+> type family BInc (binary :: Binary) :: Binary where
+>     BInc 'BEnd        = 'B1 'BEnd
+>     BInc ('B0 binary) = 'B1 binary
+>     BInc ('B1 binary) = 'B0 (BInc binary)
+
+[Appendix 3](#appendix-3) contains an (unused) implementation of incrementing
+left-to-right binary numbers.  Getting things like this to work is not too much
+of a stretch these days (even though GHCs error messages can be very cryptic).
+However, due to the large amount of type families involved, proving things about
+it presumably requires ritually sacrificing an inappropriate amount of Agda
+programmers while cantating Richard Eisenberg's writings.
+
+To that end, it is almost always worth trying to figure out alternate
+representations that work out more elegantly.  This can lead to some arbitrary
+looking choices -- we will see this in full effect when trying to define
+[CutTree](#cuttree) further below.
+
+Addition is not too hard to define:
+
+> type family BAdd (x :: Binary) (y :: Binary) :: Binary where
+>     BAdd 'BEnd   y       = y
+>     BAdd x       'BEnd   = x
+>     BAdd ('B0 x) ('B0 y) = 'B0 (BAdd x y)
+>     BAdd ('B1 x) ('B0 y) = 'B1 (BAdd x y)
+>     BAdd ('B0 x) ('B1 y) = 'B1 (BAdd x y)
+>     BAdd ('B1 x) ('B1 y) = 'B0 (BInc (BAdd x y))
+
+Let's quickly define a number of examples
+
+> type BZero  = 'B0 'BEnd
+> type BOne   = BInc BZero
+> type BTwo   = BInc BOne
+> type BThree = BInc BTwo
+> type BFour  = BInc BThree
+> type BFive  = BInc BFour
+
+This allows us to play around with it in GHCi:
+
+~~~~~
+*Main> :set -XDataKinds
+*Main> :kind! BAdd BFour BFive
+BAdd BFour BFive :: Binary
+= 'B1 ('B0 ('B0 ('B1 'BEnd)))
+~~~~~
+
+Finally, we define a corresponding singleton to use later on:
+
+> data SBin (b :: Binary) where
+>     SBEnd :: SBin 'BEnd
+>     SB0   :: SBin b -> SBin ('B0 b)
+>     SB1   :: SBin b -> SBin ('B1 b)
+
+Binomial forests
+----------------
+
+Our heap will be a relatively simple wrapper around a recursive type called
+`Forest`.  This datastructure follows the definition of the binary numbers
+fairly closely, which makes the code in this section surprisingly easy and we
+end up requiring no lemmas or proofs whatsoever here.
+
+A `Forest o b` refers to a number of trees starting with (possibly) a tree of
+order `o`.  The `b` is the binary number that indicates the shape of the forest
+-- i.e., whether we have a tree of a given order or not.
+
+Using a handwavy but convenient notation, this means that _Forest 3 101_ refers
+to binomial trees of order 3 and 5 (and no tree of order 4).
+
+> data Forest (o :: Nat) (b :: Binary) a where
+>     FEnd :: Forest o 'BEnd a
+>     F0   ::             Forest ('Succ o) b a -> Forest o ('B0 b) a
+>     F1   :: Tree o a -> Forest ('Succ o) b a -> Forest o ('B1 b) a
```

Comment:
I can use `{-# OPTIONS_GHC -Wno-unticked-promoted-constructors #-}` at the top of the file.  I'm not sure if this would make it more or less readable though.

Reply:
Seems like this was fixed in KDEs highlighters recently, which is what Pandoc uses.  Trying to update...

Reply:
Fixed